pkg/utils/utils_virtual.go

```diff
@@ -118,6 +118,24 @@ func GetOpenstackData(useHostPath bool) (metaData *OSPMetaData, networkData *OSP
 	if err != nil {
 		metaData, networkData, err = getOpenstackDataFromMetadataService()
 	}
+
+	// We can't rely on the PCI address from the metadata so we will lookup the real PCI address
+	// for the NIC that matches the MAC address.
+	//
+	// Libvirt/QEMU cannot guarantee that the address specified in the XML will match the address seen by the guest.
+	// This is a well known limitation: https://libvirt.org/pci-addresses.html
+	// When using the q35 machine type, it highlights this issue due to the change from using PCI to PCI-E bus for virtual devices.
+	//
+	// With that said, the PCI value in Nova Metadata is a best effort hint due to the limitations mentioned above. Therefore
+	// we will lookup the real PCI address for the NIC that matches the MAC address.
+	for _, device := range metaData.Devices {
+		realPCIAddr, err := getPCIAddressFromMACAddress(device.Mac)
+		if err != nil {
+			return metaData, networkData, fmt.Errorf("GetOpenstackData(): error getting PCI address for device %s: %v", device.Mac, err)
+		}
+		device.Address = realPCIAddr
```

Comment:
Rather than blindly overwriting it, could you add a debug-level warning if there's a mismatch? Eventually this will get fixed in Nova (since the fix will likely be there in both cases) and until then, any little encouragement to resolve this is a good thing.

Reply:
Good point. I'll look at that.